add author_date and committer_date revset functions

jennings opened the pull request:
<!--
There's no need to add anything here, but feel free to add a personal message.
Please describe the changes in this PR in the commit message(s) instead, with
each commit representing one logical change. Address code review comments by
rewriting the branch rather than adding commits on top. Use force-push when
pushing the updated branch (`jj git push` does that automatically when you
rewrite a branch). Merge the PR at will once it's been approved. See
https://github.com/martinvonz/jj/blob/main/docs/contributing.md for details.
Note that you need to sign Google's CLA to contribute.
-->

# Checklist

If applicable:
- [X] I have updated `CHANGELOG.md`
- [X] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [X] I have added tests to cover my changes

yuja commented on cli/src/config/revsets.toml:
```diff
@@ -21,3 +21,5 @@ latest(
 'immutable()' = '::(immutable_heads() | root())'
 'mutable()' = '~immutable()'
+
+'date(d)' = 'committer_date(d)'
```
Is this alias really needed? If user wants to compare committer timestamp, I think he can just use `committer_date()`.

jennings replied:
The only reason I added this was, while discussing the feature in Discord, some people expected the function to be `date()`. I will remove the alias though. If enough people feel it's missing, it can always be added later.

jennings replied:
Removed.

yuja commented on lib/src/revset.rs:
```diff
@@ -691,6 +697,13 @@ static BUILTIN_FUNCTION_MAP: Lazy<HashMap<&'static str, RevsetFunction>> = Lazy:
+        Ok(RevsetExpression::filter(RevsetFilterPredicate::AuthorDate(
+            pattern,
+        )))
+    });
```
It's unclear that `author_date(x)` means `t >= x`. Maybe it can be renamed to `author_date_after()` or `authored_after()`?

I also like the git-branchless' syntax. `author_date(after:x)`.
https://github.com/martinvonz/jj/issues/3287

jennings replied:
I'm open to renaming it to `X_after()`.

I originally thought the syntax was going to end up like one of these:

- `author_date(after:"2024-03-01")`
- `author_date(after="2024-03-01")`
- `author_date("after 2024-03-01")`

However, I thought this introduced a problem: It implies that `before:` should work too. And "before" has design decisions that I thought were tough to resolve in a way that satisfies most people.

- Does `author_date(before="2024")` mean `t < 2024-01-01 00:00:00` or `t < 2025-01-01 00:00:00`? The former is consistent, but a lot of people seem to expect dates to be inclusive on both ends. For example, I think a lot of people expect this to include all of March, whereas I expect it to exclude March 31: `date(after:"2024-03-01") & date(before:"2024-03-31")`

- If "before" is inclusive, then partial dates should resolve to the _end_ of their matching times:
  | revset | after | before |
  | -- | -- | -- |
  | `after:"2024"` | `after:"2024-01-01 00:00:00"` | `before:"2024-12-31 23:59:59"` |
  | `after:"2024-03-01"` | `after:"2024-03-01 00:00:00"` | `before:"2024-03-01 23:59:59"` |
  | `after:"March 1"` | `after:"2024-03-01 00:00:00"` | `before:"2024-03-01 23:59:59"` |

  chrono-english doesn't support this as far as I can tell, partial dates always resolve to the beginning of the time period. And there's no way after parsing to learn what precision was given, so we'd have to write our own parser. I originally did this with Pest, but supporting localized date formats like "April 1" vs. "1 April" seemed like it was going to be a losing battle.

Because I wasn't sure what to do, I opted to omit the word "after" entirely and just make it the default, to avoid implying that "before" should be there.

So, what do you think? Add the word "after" and then deal with feature requests for "before"?

yuja replied:
> * Does `author_date(before="2024")` mean `t < 2024-01-01 00:00:00` or `t < 2025-01-01 00:00:00`? The former is consistent, but a lot of people seem to expect dates to be inclusive on both ends. For example, I think a lot of people expect this to include all of March, whereas I expect it to exclude March 31: `date(after:"2024-03-01") & date(before:"2024-03-31")`

Appears that git-branchless do the former (@arxanas correct me if I'm wrong), but I agree I belong to the party who expects the latter "inclusive" date.

> chrono-english doesn't support this as far as I can tell, partial dates always resolve to the beginning of the time period.

Correct. There's no public API to extract inner parse result afaict.

Mercurial does brute-force parsing by trying srtftime formats in order. Maybe we can do the same without writing full-blown parser?

https://repo.mercurial-scm.org/hg/file/tip/mercurial/utils/dateutil.py#l42
https://docs.rs/chrono/latest/chrono/struct.DateTime.html#method.parse_from_str

I think it's not uncommon to look for old commits, so I expect we'll eventually want "before" query. It can be expressed as negated expression, but it has cognitive load.

emilazy replied:
I like the `before:`/`after:` syntax, in line with string patterns.

I think the exclusive interpretation of `before:"2024"` makes more sense: in English, it doesn’t make sense to say an event that happened in the middle of 2024 happened “before 2024”. My intuition expects `after="2024"` to mean 2025 onwards, too, but it’s possible that interpreting it as “after and including the beginning of 2024“ (i.e., “not before 2024”) is both more useful and easier to implement. I think that interpreting “after” that way stretches language less than an inclusive interpretation of “before” (you can imagine that `"2024"` stands for an infinitesimal instant right at the border of the start of 2024, after `2023-12-31T23:59:59.999Z` but before `2024-01-01T00:00:00.000Z`).

jennings replied:
I implemented `before:` and `after:`. I left `author_date("D")` as a shorthand for `author_date(after:"D")`.

yuja commented on lib/src/time_expression.rs:
```diff
@@ -0,0 +1,107 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(missing_docs)]
```
nit: `allow(missing_docs)` can be easily removed

jennings replied:
Docs added.

yuja commented on lib/src/time_expression.rs:
```diff
@@ -0,0 +1,107 @@
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum TimeExpression {
+    /// Represents all times at or after the given instant in time.
+    AtOrAfter(DateTime<Utc>),
```
It's probably better to use `backend::Timestamp` and `MillisSinceEpoch` in public API.

jennings replied:
Replaced `DateTime<Utc>` with `backend::Timestamp`.

yuja commented on lib/src/time_expression.rs:
nit: perhaps, this module can be merged later with `time_util` of CLI crate. Just a comment.

jennings replied:
I renamed the file to `lib/time_util.rs` so it's easier to merge `src/time_util.rs` into it later.

yuja commented on lib/src/revset.rs:
```diff
@@ -755,6 +775,22 @@ pub fn expect_string_pattern(node: &ExpressionNode) -> Result<StringPattern, Rev
+    let expression = revset_parser::expect_literal::<String>("time expression", node)?;
+    TimeExpression::parse(expression.as_str(), now).map_err(|err| {
+        RevsetParseError::expression(
+            format!("Unable to parse time expression: #{err}"),
```
nit: bad `#` char?

The error can also be attached as `.with_source(err)`.

jennings replied:
`#` removed, added `.with_source(err)`.

yuja commented on lib/src/revset.rs:
```diff
@@ -755,6 +775,22 @@ pub fn expect_string_pattern(node: &ExpressionNode) -> Result<StringPattern, Rev
     revset_parser::expect_pattern_with("string pattern", node, parse_pattern)
 }
 
+fn expect_timeexpression<Tz: TimeZone>(
```
nit: `expect_time_expression` (missing underscore)?

jennings replied:
Underscore added to all the function names I added.

yuja reviewed (commented):
Thanks!

yuja commented on lib/src/revset.rs:
```diff
@@ -755,6 +775,22 @@ pub fn expect_string_pattern(node: &ExpressionNode) -> Result<StringPattern, Rev
+    Tz::Offset: Copy,
+{
+    let expression = revset_parser::expect_literal::<String>("time expression", node)?;
+    TimeExpression::parse(expression.as_str(), now).map_err(|err| {
```
nit: `&expression` also works

jennings replied:
Thanks. I'm still very new to Rust, so I'm happy to have these nits pointed out to me so I learn to write idiomatic code.

martinvonz reviewed (commented):
> chrono-english supports US and UK dialects to disambiguate mm/dd/yy from
dd/mm/yy, but for now we default to US. This should probably be a config
setting.

I don't think it's worth a config option (but I wouldn't personally mind not having support for either little-endian or middle-endian formats for either dates or time, so maybe don't listen to me :))

emilazy commented:
I would agree that it’s preferable to only support the ISO/RFC‐style `YYYY-MM-DD` formats. The non‐big‐endian formats are too hopelessly ambiguous for communication with computers or international users. If they are unnatural to users in the major English speaking countries, they are at least comparably unnatural to all of them, and I think anyone in computing is likely to have encountered them before (if not, Jujutsu can be their introduction!).

jennings commented:
I've made the following substantive changes:

- Renamed TimeExpression to TimePattern, to fit better with StringPattern and FilePattern.
- Added an optional `after:` prefix. If no prefix is given, it behaves like `after:`.
- `before:` is now supported, and is exclusive

Since `after:` is inclusive but `before:` is exclusive, every revision falls into either `author_date(before:"D")` or `author_date(after:"D")` given any date _D_, but never both.

yuja commented on lib/src/time_pattern.rs:
```diff
@@ -0,0 +1,137 @@
+            parse_date_string(s, now, Dialect::Us).map_err(TimePatternParseError::ParseError)?;
+        let ts = Timestamp::from_datetime(d.to_utc().fixed_offset());
+        match kind {
+            None => Ok(TimePattern::AtOrAfter(ts)),
```
Can't we make this an error? I think it's confusing if `author_date("2000")` implied `>= 2000-01-01`.

jennings replied:
Will do. I was kind of attached to it as a shortcut, but that's probably just because I originally wasn't going to have `before:` at all.

yuja commented on lib/src/time_pattern.rs:
```diff
@@ -0,0 +1,137 @@
+    pub fn from_str_kind<Tz: TimeZone>(
+        s: &str,
+        kind: Option<&str>,
+        now: DateTime<Tz>,
```
nit: maybe use `backend::Timestamp` for consistency? (You might want to move time_util from cli crate.)

If this is for future DST support, some comment will help.

jennings replied:
This is for DST support, so I'll add a comment. If the local offset changed between when the command is run and the date string (like "yesterday"), the full timezone would be needed to understand what "yesterday at midnight" was.

yuja commented on CHANGELOG.md:
```diff
@@ -82,6 +82,8 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 * `jj commit` now accepts `--reset-author` option to match `jj describe`.
 
+* Added revset functions `author_date` and `committer_date`.
+
```
(reminder) you'll probably need to relocate this entry after the release.

yuja commented on lib/src/default_index/revset_engine.rs:
```diff
@@ -1069,6 +1070,30 @@ fn build_predicate_fn(
+                let entry = index.entry_by_pos(pos);
+                let commit = store.get_commit(&entry.commit_id()).unwrap();
+                let author_date = &commit.author().timestamp;
+                match &expression {
```
nit: perhaps, this can be extracted as `TimePattern::matches()` method.

yuja commented on lib/src/default_index/revset_engine.rs:
```diff
@@ -1069,6 +1070,30 @@ fn build_predicate_fn(
+                let author_date = &commit.author().timestamp;
+                match &expression {
+                    TimePattern::AtOrAfter(ts) => ts.le(author_date),
+                    TimePattern::Before(ts) => ts.gt(author_date),
```
nit: use `<=`, `>` operator instead

jennings replied:
Ah, of course.

yuja commented on lib/src/revset.rs:
```diff
@@ -131,6 +133,10 @@ pub enum RevsetFilterPredicate {
     Author(StringPattern),
     /// Commits with committer's name or email containing the needle.
     Committer(StringPattern),
+    /// Commits authored after the given date.
```
nit: "after" is no longer valid

yuja commented on lib/src/revset.rs:
```diff
@@ -1977,20 +2015,29 @@ impl RevsetExtensions {
             aliases_map,
             user_email,
+            now: now.naive_local(),
+            offset: now.offset().fix(),
```
I don't think `Tz` parameter makes sense here. If we convert `now` to `FixedOffset`, we can use `backend::Timestamp` type for API consistency.

If we want to support DST thingy, `DateTime<Local>` shouldn't be decomposed here?

jennings replied:
backend::Timestamp didn't seem appropriate because I eventually need the `DateTime<Tz>` to pass into `chrono_english::parse_date_string`. I didn't want to lose the time zone information. I resolved this by:

- Creating a DatePatternContext, which is either `Local(DateTime<Local>)` or `Utc(DateTime<Utc>)`
- Storing a DatePatternContext in the RevsetParseContext instead of any chrono types
- Call `context.to_date_pattern(s, kind)` to create the DatePattern

The production code uses `Local(DateTime<Local>)`, but tests use `Utc(DateTime<Local>)`.

yuja commented on lib/src/time_pattern.rs:
```diff
@@ -12,7 +12,7 @@
 // limitations under the License.
 
-//! Provides a TimeExpression type that represents a range of time.
+//! Provides a TimePattern type that represents a range of time.
```
nit: squash into the previous commit

yuja reviewed (approved):
Thanks!

yuja commented on lib/src/time_pattern.rs:
```diff
@@ -0,0 +1,137 @@
+    /// Unknown pattern kind is specified.
+    #[error(r#"Invalid time pattern kind "{0}:""#)]
+    InvalidKind(String),
+    /// Failed to parse input UI path.
```
nit: copy paste error: "input UI path"

yuja commented on lib/src/time_util.rs:
```diff
@@ -0,0 +1,189 @@
+
+impl DatePatternContext {
+    /// Parses a DatePattern from the given string and kind.
+    pub fn to_date_pattern(
```
nit: `to_` sounds a bit odd. Maybe this can be renamed to `parse_relative`/`_with` or just `parse`?

yuja commented on lib/src/time_util.rs:
```diff
@@ -0,0 +1,189 @@
+        s: &str,
+        kind: &str,
+    ) -> Result<DatePattern, DatePatternParseError> {
+        match &self {
```
nit: `*self` can be used here (since `DateTime<_>` can be copied.)

yuja commented on lib/src/time_util.rs:
```diff
@@ -0,0 +1,189 @@
+    {
+        let d =
+            parse_date_string(s, now, Dialect::Us).map_err(DatePatternParseError::ParseError)?;
+        let ts = Timestamp::from_datetime(d.to_utc().fixed_offset());
```
Maybe `.to_utc()` is unneeded?

We can also change the underlying type to `MillisSinceEpoch` to clarify that the tz offset doesn't matter.

yuja commented on lib/tests/test_revset.rs:
```diff
@@ -2426,6 +2437,144 @@ fn test_evaluate_expression_author() {
+        vec![commit1.id().clone(), root_commit.id().clone()]
+    );
+}
+
```
Can you add a few `Local` timezone tests to exercise the production code? It could be a CLI tests.

jennings replied:
Added a test `test_revset_committer_date_with_timezone` that creates some commits and verifies that, after changing the TZ env var, the matching revisions changes.

It would probably be nice to verify that it understands DST shifts, but I'm not sure how to write that test because I'd need to control the current time. For example, if it's 2024-06-01 and UTC-4 in New York and you enter `committer_date(after:"2024-01-01")`, it _should_ know that the offset at that time was UTC-5 so that means `2024-01-01T05:00:00Z`.

yuja replied:
Maybe we can test with two or three different timezones whose DST ranges don't overlap? I don't know how hard it would be, though.

BTW, chrono appears to support `TZ` variable with DST range:
https://github.com/chronotope/chrono/blob/43579a5304f9433ce42614e200b554ea8e0361cc/src/offset/local/tz_info/rule.rs#L20
It'll probably help stabilize the tests.

jennings replied:
I added another check that I get the expected results from an Australian time zone. I feel pretty confident that it's behaving as I wanted.

yuja commented on lib/src/revset.rs:
```diff
@@ -752,6 +771,22 @@ pub fn expect_string_pattern(node: &ExpressionNode) -> Result<StringPattern, Rev
+        )),
+        Some(kind) => context.to_date_pattern(value, kind).map_err(|err| {
+            RevsetParseError::expression(format!("Unable to parse date pattern: {err}"), node.span)
+                .with_source(err)
```
This duplicates error messages. We have to manually convert the error to `Box<dyn _>`.
```rust
        |value, kind| -> Result<_, Box<dyn std::error::Error + Send + Sync>> {
            match kind {
                None => Err("Date pattern must specify 'after' or 'before'".into()),
                Some(kind) => Ok(context.to_date_pattern(value, kind)?),
            }
        }
```

jennings commented on lib/src/time_util.rs:
```diff
@@ -0,0 +1,189 @@
+    /// Interpret date patterns using the local machine's time zone
+    Local(DateTime<Local>),
+    /// Interpret date patterns using UTC
+    Utc(DateTime<Utc>),
```
Added.

yuja replied:
Oops, I meant `Utc` can be replaced with `Fixed`. It's okay to have these three though.

yuja commented on lib/src/time_util.rs:
```diff
@@ -0,0 +1,191 @@
+        match *self {
+            DatePatternContext::Local(dt) => DatePattern::from_str_kind(s, kind, dt.to_owned()),
+            DatePatternContext::Utc(dt) => DatePattern::from_str_kind(s, kind, dt.to_owned()),
+            DatePatternContext::Fixed(dt) => DatePattern::from_str_kind(s, kind, dt.to_owned()),
```
nit: `.to_owned()` is now redundant

jennings replied:
Thanks. I'm still trying to internalize the details of Rust ownership.

I thought I needed `.to_owned()` because `from_str_kind` takes a `DateTime<Tz>`, not `&DateTime<Tz>`, and Tz may not be Copy. But I think I don't anymore because `DateTime<Local>`, `DateTime<Utc>`, and `DateTime<FixedOffset>` are all Copy, so the compiler knows it can make a new owned value by just copying bits. Am I understanding correctly?

yuja replied:
They're all `Copy`, so `*self` (where: `self: &DatePatternContext`) can make a copy of `dt: &DateTime<_>`, and now `dt: DateTime<_>`. Another option is `*dt` (without `*self`.)

yuja commented on lib/src/time_util.rs:
```diff
@@ -0,0 +1,191 @@
+
+#[cfg(test)]
+mod tests {
+    use chrono::DateTime;
```
nit: redundant `use `
